File: reader/server/core/backend/src/main/java/com/richodemus/reader/backend/Backend.java
package com.richodemus.reader.backend;

import com.richodemus.reader.backend.exception.ItemNotInFeedException;
import com.richodemus.reader.backend.exception.NoSuchChannelException;
import com.richodemus.reader.backend.exception.NoSuchUserException;
import com.richodemus.reader.backend.exception.UserNotSubscribedToThatChannelException;
import com.richodemus.reader.backend.feed.FeedRepository;
import com.richodemus.reader.backend.model.Feed;
import com.richodemus.reader.backend.model.FeedWithoutItems;
import com.richodemus.reader.backend.model.Item;
import com.richodemus.reader.backend.subscription.SubscriptionRepository;
import com.richodemus.reader.backend.user.UserRepository;
import com.richodemus.reader.dto.FeedId;
import com.richodemus.reader.dto.FeedName;
import com.richodemus.reader.dto.FeedUrl;
import com.richodemus.reader.dto.ItemId;
import com.richodemus.reader.dto.UserId;
import com.richodemus.reader.dto.Username;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;


import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Timer;

import static java.util.Collections.emptyList;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;

@Component
public class Backend
{
	private final Logger logger = LoggerFactory.getLogger(getClass());
	private final SubscriptionRepository subscriptionRepository;
	private final FeedRepository feedRepository;
	private final UserRepository userRepository;

	public Backend(final SubscriptionRepository subscriptionRepository,
				   final FeedRepository feedRepository,
				   final UserRepository userRepository)
	{
		this.subscriptionRepository = subscriptionRepository;
		this.feedRepository = feedRepository;
		this.userRepository = userRepository;

	}

	public Optional<Feed> getFeed(Username username, FeedId feedId)
	{
		logger.debug("Getting feed {} for user {}", feedId, username);
		final UserId userId = userRepository.getUserId(username);
		final Optional<Feed> feed = feedRepository.getFeed(feedId);
		if (!feed.isPresent())
		{
			logger.warn("No such feed {}", feedId);
			return Optional.empty();
		}

		final List<ItemId> watchedItems = subscriptionRepository.get(userId, feedId);
		final List<Item> items = feed.get().getItems().stream().filter(item -> !watchedItems.contains(item.getId())).collect(toList());

		return Optional.of(new Feed(feedId, feed.get().getName(), items));
	}

	public List<FeedWithoutItems> getAllFeedsWithoutItems(Username username)
	{
		logger.debug("Getting all feeds for user {}", username);
		final UserId userId = userRepository.getUserId(username);

		final List<Feed> feeds = subscriptionRepository.get(userId);

		final Map<FeedId, Feed> feedIds = getAllFeedsForUser(feeds);

		return mergeFeeds(feeds, feedIds);

	}

	private Map<FeedId, Feed> getAllFeedsForUser(List<Feed> feeds)
	{
		return feeds.stream()
				.collect(toMap(Feed::getId, feed ->
				{
					final Optional<Feed> feedRepositoryFeed = feedRepository.getFeed(feed.getId());
					return feedRepositoryFeed.orElseGet(() -> new Feed(feed.getId(), new FeedName("UNKNOWN_FEED"), emptyList()));
				}));
	}

	private List<FeedWithoutItems> mergeFeeds(List<Feed> feeds, Map<FeedId, Feed> feedsWithItems)
	{
		List<FeedWithoutItems> results = new ArrayList<>();

		for (Feed feed : feeds)
		{
			final List<Item> watchedItems = new ArrayList<>(feed.getItems());
			final Feed repoFeed = feedsWithItems.get(feed.getId());
			final List<Item> itemIds = repoFeed.getItems();

			int numberOfUnwatchedItems = itemIds.size();

			for (Item item : itemIds)
			{
				for (int i = 0; i < watchedItems.size(); i++)
				{

					if (item.getId().equals(watchedItems.get(i).getId()))
					{
						numberOfUnwatchedItems--;
						watchedItems.remove(i);
						break;
					}
				}
			}

			results.add(new FeedWithoutItems(feed.getId(), repoFeed.getName(), numberOfUnwatchedItems));
		}

		return results;
	}

	public void addFeed(final Username username, final FeedUrl feedUrl) throws NoSuchChannelException, NoSuchUserException
	{
		logger.info("Add feed: {} for user {}", feedUrl, username);

		final FeedId feedId = feedRepository.getFeedId(feedUrl);
		final UserId userId = userRepository.getUserId(username);

		//Todo its now possible to add feeds that doesnt exist...
		subscriptionRepository.subscribe(userId, feedId);
	}

	public void markAsRead(final Username username, final FeedId feedId, final ItemId itemId) throws NoSuchUserException, UserNotSubscribedToThatChannelException
	{
		logger.info("Marking item {} in feed {} for user {} as read", itemId, feedId, username);
		final UserId userId = userRepository.getUserId(username);
		subscriptionRepository.markAsRead(userId, feedId, itemId);
	}

	public void markAsUnread(final Username username, final FeedId feedId, ItemId itemId) throws NoSuchUserException
	{
		logger.info("Marking item {} in feed {} for user {} as unread", itemId, feedId, username);
		final UserId userId = userRepository.getUserId(username);
		subscriptionRepository.markAsUnread(userId, feedId, itemId);
	}

	public void markOlderItemsAsRead(final Username username, final FeedId feedId, final ItemId itemId) throws NoSuchChannelException, ItemNotInFeedException, UserNotSubscribedToThatChannelException
	{
		logger.info("Marking items older than {} in feed {} for user {} as read", itemId, feedId, username);
		final UserId userId = userRepository.getUserId(username);

		final Feed feed = feedRepository.getFeed(feedId).orElseThrow(() ->
		{
			logger.error("No such channel: {}", feedId);
			return new NoSuchChannelException("No such channel: " + feedId);
		});

		final Item targetItem = feed.getItems().stream()
				.filter(item -> item.getId().equals(itemId))
				.findAny()
				.orElseThrow(() -> new ItemNotInFeedException("Item " + itemId + " is not in feed " + feedId));

		feed.getItems().stream()
				.filter(item -> item.isBefore(targetItem))
				.map(Item::getId)
				.forEach(id -> subscriptionRepository.markAsRead(userId, feedId, id));
	}
}

File: reader/server/core/backend/src/main/kotlin/com/richodemus/reader/backend/Backend.kt
package com.richodemus.reader.backend

import com.richodemus.reader.backend.exception.ItemNotInFeedException
import com.richodemus.reader.backend.exception.NoSuchChannelException
import com.richodemus.reader.backend.exception.NoSuchUserException
import com.richodemus.reader.backend.exception.UserNotSubscribedToThatChannelException
import com.richodemus.reader.backend.feed.FeedRepository
import com.richodemus.reader.backend.model.Feed
import com.richodemus.reader.backend.model.FeedWithoutItems
import com.richodemus.reader.backend.model.Item
import com.richodemus.reader.backend.subscription.SubscriptionRepository
import com.richodemus.reader.backend.user.UserRepository
import com.richodemus.reader.dto.FeedId
import com.richodemus.reader.dto.FeedName
import com.richodemus.reader.dto.FeedUrl
import com.richodemus.reader.dto.ItemId
import com.richodemus.reader.dto.Username
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.stream.Collectors

@Component
class Backend(private val subscriptionRepository: SubscriptionRepository,
              private val feedRepository: FeedRepository,
              private val userRepository: UserRepository) {
    private val logger = LoggerFactory.getLogger(javaClass)
    fun getFeed(username: Username, feedId: FeedId): Feed? {
        logger.debug("Getting feed {} for user {}", feedId, username)
        val userId = userRepository.getUserId(username)
        val feed = feedRepository.getFeed(feedId)
        if (feed == null) {
            logger.warn("No such feed {}", feedId)
            return null
        }
        val watchedItems = subscriptionRepository[userId, feedId]
        val items = feed.items.stream().filter { item: Item -> !watchedItems.contains(item.id) }.collect(Collectors.toList())
        return Feed(feedId, feed.name, items)
    }

    fun getAllFeedsWithoutItems(username: Username): List<FeedWithoutItems> {
        logger.debug("Getting all feeds for user {}", username)
        val userId = userRepository.getUserId(username)
        val feeds = subscriptionRepository[userId]
        val feedIds = getAllFeedsForUser(feeds)
        return mergeFeeds(feeds, feedIds)
    }

    private fun getAllFeedsForUser(feeds: List<Feed>): Map<FeedId, Feed> {
        return feeds.stream()
                .collect(Collectors.toMap({ obj: Feed -> obj.id }) { feed: Feed ->
                    val feedRepositoryFeed = feedRepository.getFeed(feed.id)
                    feedRepositoryFeed ?: Feed(feed.id, FeedName("UNKNOWN_FEED"), emptyList())
                })
    }

    private fun mergeFeeds(feeds: List<Feed>, feedsWithItems: Map<FeedId, Feed>): List<FeedWithoutItems> {
        val results: MutableList<FeedWithoutItems> = ArrayList()
        for (feed in feeds) {
            val watchedItems = ArrayList(feed.items)
            val repoFeed = feedsWithItems[feed.id]
            val itemIds = repoFeed?.items
            var numberOfUnwatchedItems = itemIds?.size ?: 0
            if (itemIds != null) {
                for (item in itemIds) {
                    for (i in watchedItems.indices) {
                        if (item.id == watchedItems[i]?.id) {
                            numberOfUnwatchedItems--
                            watchedItems.removeAt(i)
                            break
                        }
                    }
                }
            }
            if (repoFeed != null) {
                results.add(FeedWithoutItems(feed.id, repoFeed.name, numberOfUnwatchedItems))
            }
        }
        return results
    }

    @Throws(NoSuchChannelException::class, NoSuchUserException::class)
    fun addFeed(username: Username, feedUrl: FeedUrl) {
        logger.info("Add feed: {} for user {}", feedUrl, username)
        val feedId = feedRepository.getFeedId(feedUrl)
        val userId = userRepository.getUserId(username)

        //Todo its now possible to add feeds that doesnt exist...
        subscriptionRepository.subscribe(userId, feedId!!)
    }

    @Throws(NoSuchUserException::class, UserNotSubscribedToThatChannelException::class)
    fun markAsRead(username: Username, feedId: FeedId, itemId: ItemId) {
        logger.info("Marking item {} in feed {} for user {} as read", itemId, feedId, username)
        val userId = userRepository.getUserId(username)
        subscriptionRepository.markAsRead(userId, feedId, itemId)
    }

    @Throws(NoSuchUserException::class)
    fun markAsUnread(username: Username, feedId: FeedId, itemId: ItemId) {
        logger.info("Marking item {} in feed {} for user {} as unread", itemId, feedId, username)
        val userId = userRepository.getUserId(username)
        subscriptionRepository.markAsUnread(userId, feedId, itemId)
    }

    @Throws(NoSuchChannelException::class, ItemNotInFeedException::class, UserNotSubscribedToThatChannelException::class)
    fun markOlderItemsAsRead(username: Username, feedId: FeedId, itemId: ItemId) {
        logger.info("Marking items older than {} in feed {} for user {} as read", itemId, feedId, username)
        val userId = userRepository.getUserId(username)
        val feed = feedRepository.getFeed(feedId)
        if (feed == null) {
            logger.error("No such channel: {}", feedId)
            throw NoSuchChannelException("No such channel: $feedId")
        }
        val targetItem = feed.items.stream()
                .filter { item: Item -> item.id == itemId }
                .findAny()
                .orElseThrow { ItemNotInFeedException("Item $itemId is not in feed $feedId") }
        feed.items.stream()
                .filter { item: Item -> item.isBefore(targetItem) }
                .map { obj: Item -> obj.id }
                .forEach { id: ItemId -> subscriptionRepository.markAsRead(userId, feedId, id) }
    }
}
